packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -541,9 +540,8 @@ InternalModel.prototype = {
   },
 
   setId: function(id) {
+    Ember.assert('Object is a newly created model instance', this.isNew);
```

Comment:
Also lets make the assertion message say something like "A record's id can not be chanced once it is in the loaded state."